components/tikv_util/src/quota_limiter.rs

```diff
@@ -0,0 +1,202 @@
+            write_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            read_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            req_rate_limiter: Limiter::new(f64::INFINITY),
+        }
+    }
+}
+
+impl QuotaLimiter {
+    // 1000 millicpu equals to 1vCPU, 0 means unlimited
+    pub fn new(
+        cpu_quota: usize,
+        req_rate: usize,
+        write_kvs: usize,
+        write_bandwidth: ReadableSize,
+        read_bandwidth: ReadableSize,
+    ) -> Self {
+        let cputime_limiter = if cpu_quota == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            // The cpu time is not a real statistics, only part of the processing logic is
+            // taken into account, so it needs to be multiplied by a factor.
+            // transfer milli cpu to micro cpu
+            Limiter::new(cpu_quota as f64 * CPU_TIME_FACTOR * 1000_f64)
+        };
+
+        let write_kvs_limiter = if write_kvs == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_kvs as f64)
+        };
+
+        let write_bandwidth_limiter = if write_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_bandwidth.0 as f64)
+        };
+
+        let read_bandwidth_limiter = if read_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(read_bandwidth.0 as f64)
+        };
+
+        let total_qps_limiter = if req_rate == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(req_rate as f64)
+        };
+
+        Self {
+            cputime_limiter,
+            write_kvs_limiter,
+            write_bandwidth_limiter,
+            read_bandwidth_limiter,
+            req_rate_limiter: total_qps_limiter,
+        }
+    }
+
+    pub fn consume_write(
+        &self,
+        req_cnt: usize,
+        kv_cnt: usize,
+        bytes: usize,
+        cpu_micro_secs: usize,
+    ) -> Duration {
+        let cpu_dur = self.cputime_limiter.consume_duration(cpu_micro_secs);
+
+        let kv_dur = if kv_cnt > 0 {
+            self.write_kvs_limiter.consume_duration(kv_cnt)
+        } else {
+            Duration::ZERO
+        };
+
+        let bw_dur = if bytes > 0 {
+            self.write_bandwidth_limiter.consume_duration(bytes)
+        } else {
+            Duration::ZERO
+        };
+
+        let req_rate_dur = if req_cnt > 0 {
+            self.req_rate_limiter.consume_duration(req_cnt)
+        } else {
+            Duration::ZERO
+        };
+
+        std::cmp::max(
+            req_rate_dur,
+            std::cmp::max(std::cmp::max(cpu_dur, kv_dur), bw_dur),
+        )
+    }
+
+    pub fn consume_read(
+        &self,
+        req_cnt: usize,
+        read_bytes: usize,
+        cpu_micro_secs: usize,
+    ) -> Duration {
+        let cpu_dur = self.cputime_limiter.consume_duration(cpu_micro_secs);
```

Comment:
add check to totally skip consume_duration for better performance when cpu limiter is infinity.

same as other limiters.

Reply:
same in https://github.com/tikv/tikv/pull/12151#issuecomment-1066823807